### PR TITLE
Evaluation pipeline

### PR DESCRIPTION
**Pull request by rhadi-moz**
# Overview 
Evaluation pipeline for Python based retrieval for search your history POC 

Contains: 
- retrieval.py (performs retrieval) 
- evaluation.py (given a file of retrieved results, performs traditional and LLM based evaluation) 
- src/llm_judge (class for LLM evaluation, recommended to run on GPU) 
- readme (how to run) 
- requirements.txt 

# Structure 
I initially had retrieval + evaluation in the same script, but decided to separate it out to allow easier evaluation of browser based retrieval  

# Things to look at 
- Metric calculations in evaluation.py

### CODE REVIEW

File: evaluation_pipeline/evaluation.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+    Returns:
+        float: Recall@K score.
+    """
+    retrieved_at_k = retrieved_docs[:k]
+    intersection = set(retrieved_at_k) & set(relevant_docs)
+    return len(intersection) / len(relevant_docs)
+
+
+def calc_ndcg(relevant_docs, retrieved_docs, k=None):
+    """
```
Comment (chidamnat) [inline evaluation_pipeline/evaluation.py:67]: recommend reusing from sklearn - https://scikit-learn.org/stable/modules/generated/sklearn.metrics.ndcg_score.html

File: evaluation_pipeline/evaluation.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+    Compute Mean Reciprocal Rank (MRR).
+    Parameters:
+        relevant_docs (set): Set of relevant document IDs.
+        retrieved_docs (list): List of retrieved document IDs in ranked order.
+    Returns:
+        float: MRR score.
+    """
+    for rank, doc in enumerate(retrieved_docs, start=1):
+        if doc in relevant_docs:
+            return 1 / rank
```
Comment (chidamnat) [inline evaluation_pipeline/evaluation.py:98]: could not find the aggregated mean

File: evaluation_pipeline/evaluation.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+    """
+    relevant_retrieved = 0
+    precision_sum = 0.0
+
+    for k, doc in enumerate(retrieved_docs, start=1):
+        if doc in relevant_docs:
+            relevant_retrieved += 1
+            precision_sum += relevant_retrieved / k
+
+    return precision_sum / len(relevant_docs) if relevant_docs else 0.0
```
Comment (chidamnat) [inline evaluation_pipeline/evaluation.py:119]: trying sklearn version here.
Example: we have 4 docs [1,2,3,4]
ground truth [3,4]
predicted docs in ranking order [4]

```
import numpy as np
from sklearn.metrics import average_precision_score
y_true = np.array([0, 0, 1, 1]).   # ground truth
y_scores = np.array([0, 0, 0, 1]) # predicted docs in ranking order
average_precision_score(y_true, y_scores)
```
we get `np.float64(0.75)`

whereas trying calc_mean_average_precision(set([3,4]), [4])
got `0.5`

could we try to see if we could re-use https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html

File: evaluation_pipeline/retrieval.py [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+
+# Configure logging
+logging.basicConfig(
+    filename="performance.log", 
+    level=logging.INFO,
+    format="%(asctime)s - %(message)s"
+)
+
+
+def log_performance(func):
```
Comment (chidamnat) [inline evaluation_pipeline/retrieval.py:41]: nice

File: evaluation_pipeline/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,173 @@
+aiohappyeyeballs==2.4.4
```
Comment (chidamnat) [inline evaluation_pipeline/requirements.txt:1]: seems to be a lot of libraries in the requirements.txt. do you see any issues if we keep the main highlevel packages instead ?

File: evaluation_pipeline/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+aiohttp==3.11.11
```
Comment (chidamnat) [inline evaluation_pipeline/requirements.txt:2]: while running evaluation, noticed `python-dotenv` may be missing here.

File: evaluation_pipeline/evaluation.py [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+    for df, label in zip(dfs_to_return, df_labels):
+        summary_df = df.describe().reset_index()
+        summary_df = summary_df.loc[:, summary_df.columns != 'query_id'] if 'query_id' in summary_df.columns else summary_df
+        summary_table = wandb.Table(dataframe=summary_df)
+        averages = summary_df[summary_df['index'] == 'mean']
+        if log_to_wandb:
+            wandb.log({label: summary_table})
+            wandb.log({"Averages": averages})
+            wandb.finish()
+        summary_df.to_csv(f"{model_name}_{label}_aggregate_metrics.csv")
```
Comment (chidamnat) [inline evaluation_pipeline/evaluation.py:287]: could we please write these two files into results/ directory

**Review by chidamnat — APPROVED** [review]
few comments. overall looks great and was able to run this end to end.

**Comment by chidamnat** [comment]
The history extraction query could be better of with some of the conditions and sort order as below.
```
SELECT url_hash, url, title, description
       FROM places.moz_places
       WHERE description NOTNULL
       AND title NOTNULL
       ORDER BY frecency DESC
       LIMIT <rowLimit>
```

File: evaluation_pipeline/evaluation.py (hunk 7 above)
Comment (rhadi-moz) [reply evaluation_pipeline/evaluation.py:287]: Absolutely!

File: evaluation_pipeline/requirements.txt (hunk 6 above)
Comment (rhadi-moz) [reply evaluation_pipeline/requirements.txt:2]: Great catch, will fix

File: evaluation_pipeline/requirements.txt (hunk 5 above)
Comment (rhadi-moz) [reply evaluation_pipeline/requirements.txt:1]: Will fix using pipreqs tool

File: evaluation_pipeline/evaluation.py (hunk 2 above)
Comment (rhadi-moz) [reply evaluation_pipeline/evaluation.py:98]: This is for each individual query, the mean would get calculated when we take the average across all the queries (around line 293, running the describe().  I will update the naming for this function to be reciprocal rank instead of mean reciprocal rank

File: evaluation_pipeline/evaluation.py (hunk 1 above)
Comment (rhadi-moz) [reply evaluation_pipeline/evaluation.py:67]: I started down the sklearn route but then shifted away from it when the results were lists/sets in lieu of actual scores/probabilities. 

What do you think about something like this? 

```
def calc_ndcg(relevant_docs, retrieved_docs, score_type='rank', retrieved_distances=None, k=None):
    y_true = np.array([[1 if doc in relevant_docs else 0 for doc in retrieved_docs]])
    if score_type == 'rank':
        y_scores = np.array([[1 / (rank + 1) for rank in range(len(retrieved_docs))]])
    elif score_type == 'distance':
        y_scores = np.array([[1 - dist for dist in retrieved_distances]])
    ndcg = ndcg_score(y_true, y_scores)
    return ndcg 
```

File: evaluation_pipeline/evaluation.py (hunk 3 above)
Comment (rhadi-moz) [reply evaluation_pipeline/evaluation.py:119]: Renamed to "average precision" because I was calculating mean average precision too early (this is at a single query level, the mean average precision is across all queries in the run)

Comment (rhadi-moz) [reply evaluation_pipeline/evaluation.py:119]: Aside from that, focusing on the "average precision" part,  I'm running into a similar challenge when it comes to how we construct y_true & y_scores based on retrieved_docs and relevant_docs, mainly, should y_true have the same length as retrieved_docs?  

Check out: 
https://builtin.com/articles/mean-average-precision

Will grab some time to discuss!

Comment (chidamnat) [reply evaluation_pipeline/evaluation.py]: please ignore sklearn suggestion. that is not the real MAP calculation. 
Please find the test case as below to validate.

calc_map_at_k([3,4], [3,5], k=2)
```
[1.0, 0.5]    # precision at each position
0.75           # avg precision for this query
```

example 2:
```
actuals = [1, 3, 6, 9, 10]
preds = [1,2,3,4,5]
calc_map_at_k(actuals, preds, k=5)
```
```
[1.0, 0.5, 0.6666666666666666, 0.5, 0.4]
0.6133333333333333
```

Comment (rhadi-moz) [reply evaluation_pipeline/evaluation.py]: Validating test cases! (didn't print the precision in my final function, but validated the list level precision as a part of exercise) 

image
 
`
def calc_average_precision(relevant_docs, retrieved_docs, k):
    """
    Compute Average Precision (AP).
    Parameters:
        relevant_docs (set): Set of relevant document IDs.
        retrieved_docs (list): List of retrieved document IDs in ranked order.
        k: average precision from 1 to k 
    Returns:
        float: AP score.
        Can be used to calculate mean average precision for number of queries Q
    """
    if len(retrieved_docs) < 2:
        return 0.0
    total_precision = 0.0 
    for i in range(1, len(retrieved_docs)+1): 
        precision = calc_precision_at_k(relevant_docs, retrieved_docs, k=i)
        total_precision += precision 
    if k > len(retrieved_docs): 
        print(f"k is higher than retrieval {len(retrieved_docs)}") 
    elif k < len(retrieved_docs): 
        print(f"k is lower than retrieval {len(retrieved_docs)}") 
    average_precision = total_precision / k 
    return average_precision

`
